### PR TITLE
[NOTEPAD] Improve UTF-8 encoding support and improve status bar

### PR DESCRIPTION
**Pull request by TAN-Gaming**
## Proposed changes
- Add write support for UTF-8 without BOM
- Show line endings and encoding on the status bar (like Notepad in Windows 10)

ROS_Notepad_New3

### CODE REVIEW

**Comment by HBelusca** [comment]
Depending on the amount of code that was originated from Wine, it might be of interest to send them a patch.

File: base/applications/notepad/main.c [hunk 1]
```diff
@@ -427,8 +427,13 @@ NOTEPAD_WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam)
...
 
-            if (hdwp != NULL)
-                EndDeferWindowPos(hdwp);
+            if (hdwp == NULL)
+                break;
+                
+            EndDeferWindowPos(hdwp);
+
+            /* Align status bar parts */
+            DIALOG_StatusBarAlignParts();
```
Comment (HBelusca) [inline base/applications/notepad/main.c:436]: You sure you don't want to do alignment if the hdwp operations above failed at some point?

File: base/applications/notepad/dialog.c [hunk 2]
```diff
@@ -104,6 +118,56 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+
+    if (parts[0] < defaultWidths[0])
+    {
+        parts[0] = defaultWidths[0];
+    }
+
+    if (parts[1] < defaultWidths[0] + defaultWidths[1])
+    {
+        parts[1] = defaultWidths[0] + defaultWidths[1];
+    }
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:141]:
```suggestion
    parts[0] = max(parts[0], defaultWidths[0]);
    parts[0] = max(parts[1], defaultWidths[0] + defaultWidths[1]);
```

File: base/applications/notepad/dialog.c [hunk 3]
```diff
@@ -104,6 +118,56 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+    WCHAR szText[128];
+
+    LoadStringW(Globals.hInstance, EolnToStrId[Globals.iEoln], szText, _countof(szText));
+
+    SendMessageW(Globals.hStatusBar, SB_SETTEXTW, 1, (LPARAM)szText);
+}
+
+static VOID DIALOG_StatusBarUpdateEncoding(VOID)
+{
+    WCHAR szText[128];
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:157]:
```suggestion
    WCHAR szText[128] = L"";
```

File: base/applications/notepad/dialog.c [hunk 4]
```diff
@@ -104,6 +118,56 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+    WCHAR szText[128];   
+
+    if (Globals.encFile != ENCODING_AUTO)
+    {
+        LoadStringW(Globals.hInstance, EncToStrId[Globals.encFile], szText, _countof(szText));
+    }
+    else
+    {
+        szText[0] = UNICODE_NULL;
+    }
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:166]:
```suggestion
```

File: base/applications/notepad/notepad.h [hunk 5]
```diff
@@ -16,6 +16,7 @@
 #include <commdlg.h>
 #include <tchar.h>
 #include <malloc.h>
+#include <stdlib.h>
```
Comment (HBelusca) [inline base/applications/notepad/notepad.h:19]: For which new functionality do you need this header?

File: base/applications/notepad/dialog.c (hunk 4 above)
Comment (TAN-Gaming) [reply base/applications/notepad/dialog.c]: Done.

File: base/applications/notepad/dialog.c (hunk 3 above)
Comment (TAN-Gaming) [reply base/applications/notepad/dialog.c]: Done.

File: base/applications/notepad/dialog.c (hunk 2 above)
Comment (TAN-Gaming) [reply base/applications/notepad/dialog.c]: Done.

File: base/applications/notepad/notepad.h (hunk 5 above)
Comment (TAN-Gaming) [reply base/applications/notepad/notepad.h:19]: For `_countof` macro. Or should I use `ARRAY_SIZE` instead?

File: base/applications/notepad/main.c (hunk 1 above)
Comment (TAN-Gaming) [reply base/applications/notepad/main.c:436]: Yes, the alignment is done only if the resize operation of the status bar succeeds.

File: base/applications/notepad/notepad.h (hunk 5 above)
Comment (katahiromz) [reply base/applications/notepad/notepad.h:19]: If `_countof` available, use it.

File: base/applications/notepad/main.c [hunk 6]
```diff
@@ -427,8 +427,13 @@ NOTEPAD_WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam)
...
                                   SWP_NOZORDER);
 
-            if (hdwp != NULL)
-                EndDeferWindowPos(hdwp);
+            if (hdwp == NULL)
+                break;
+                
+            EndDeferWindowPos(hdwp);
+
+            /* Align status bar parts */
```
Comment (HBelusca) [inline base/applications/notepad/main.c:435]:
```suggestion
            /* Align status bar parts, only if the status bar resize operation succeeds */
```

File: base/applications/notepad/dialog.c [hunk 7]
```diff
@@ -104,6 +125,47 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+    GetClientRect(Globals.hStatusBar, &rcStatusBar);
+
+    parts[0] = rcStatusBar.right - (defaultWidths[1] + defaultWidths[2]);
+    parts[1] = rcStatusBar.right - defaultWidths[2];
+    parts[2] = -1; // the right edge of the status bar
+
+    parts[0] = max(parts[0], defaultWidths[0]);
+    parts[1] = max(parts[1], defaultWidths[0] + defaultWidths[1]);
+
+    SendMessageW(Globals.hStatusBar, SB_SETPARTS, (WPARAM)_countof(parts), (LPARAM)parts);
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:143]:
```suggestion
    SendMessageW(Globals.hStatusBar, SB_SETPARTS, (WPARAM)ARRAY_SIZE(parts), (LPARAM)parts);
```

File: base/applications/notepad/dialog.c [hunk 8]
```diff
@@ -104,6 +125,47 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+}
+
+static VOID DIALOG_StatusBarUpdateLineEndings(VOID)
+{
+    WCHAR szText[128];
+
+    LoadStringW(Globals.hInstance, EolnToStrId[Globals.iEoln], szText, _countof(szText));
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:150]:
```suggestion
    LoadStringW(Globals.hInstance, EolnToStrId[Globals.iEoln], szText, ARRAY_SIZE(szText));
```

File: base/applications/notepad/dialog.c [hunk 9]
```diff
@@ -104,6 +125,47 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+
+static VOID DIALOG_StatusBarUpdateEncoding(VOID)
+{
+    WCHAR szText[128];
+
+    szText[0] = UNICODE_NULL;
+
+    if (Globals.encFile != ENCODING_AUTO)
+    {
+        LoadStringW(Globals.hInstance, EncToStrId[Globals.encFile], szText, _countof(szText));
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:163]:
```suggestion
        LoadStringW(Globals.hInstance, EncToStrId[Globals.encFile], szText, ARRAY_SIZE(szText));
```

File: base/applications/notepad/notepad.h (hunk 5 above)
Comment (HBelusca) [inline base/applications/notepad/notepad.h:19]:
```suggestion
```

File: base/applications/notepad/dialog.c [hunk 10]
```diff
@@ -104,6 +125,47 @@ void UpdateWindowCaption(BOOL clearModifyAlert)
...
+    LoadStringW(Globals.hInstance, EolnToStrId[Globals.iEoln], szText, _countof(szText));
+
+    SendMessageW(Globals.hStatusBar, SB_SETTEXTW, SBPART_EOLN, (LPARAM)szText);
+}
+
+static VOID DIALOG_StatusBarUpdateEncoding(VOID)
+{
+    WCHAR szText[128];
+
+    szText[0] = UNICODE_NULL;
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:159]:
```suggestion
    WCHAR szText[128] = L"";
```

File: base/applications/notepad/dialog.c [hunk 11]
```diff
@@ -403,6 +465,10 @@ VOID DoOpenFile(LPCTSTR szFileName)
     SetFileName(szFileName);
     UpdateWindowCaption(TRUE);
     NOTEPAD_EnableSearchMenu();
+
+    /* Update line endings and encoding on the status bar */
+    DIALOG_StatusBarUpdateLineEndings();
+    DIALOG_StatusBarUpdateEncoding();
```
Comment (HBelusca) [inline base/applications/notepad/dialog.c:471]:
```suggestion
    DIALOG_StatusBarUpdateEncoding();

```

**Review by HBelusca — APPROVED** [review]
OK

**Comment by JoachimHenze** [comment]
The thing with the size-grip was hackfixed by this PR just. That was no proper fix, see https://jira.reactos.org/browse/CORE-16528?focusedCommentId=138189&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-138189

**Pull request merged**